File: market/precision/precision.go
// Package precision provides the fixed-point arithmetic used by authoritative
// market calculations. Values are scaled integers; callers must not use
// binary floating point for settlement inputs or outputs.
package precision

import (
	"errors"
	"math/big"
)

const (
	// WeightScale matches the existing immutable Keeper catalog representation.
	WeightScale int64 = 100_000_000
	// ReturnScale stores component and basket returns to eight decimal places.
	ReturnScale int64 = 100_000_000
	// NormalizedScale stores normalized performance and percentile values.
	NormalizedScale int64 = 1_000_000
	// ScoreScale stores score components to four decimal places.
	ScoreScale int64 = 10_000
)

var (
	ErrDivisionByZero = errors.New("precision division by zero")
	ErrOverflow       = errors.New("precision overflow")
	ErrInvalidRange   = errors.New("precision invalid clamp range")
)

// Add returns left + right when the exact result fits in int64.
func Add(left, right int64) (int64, error) {
	return checkedBigInt(new(big.Int).Add(big.NewInt(left), big.NewInt(right)))
}

// Subtract returns left - right when the exact result fits in int64.
func Subtract(left, right int64) (int64, error) {
	return checkedBigInt(new(big.Int).Sub(big.NewInt(left), big.NewInt(right)))
}

// MultiplyDivide calculates left*right/divisor, rounding half away from zero.
// Intermediate multiplication is exact and cannot overflow int64.
func MultiplyDivide(left, right, divisor int64) (int64, error) {
	if divisor == 0 {
		return 0, ErrDivisionByZero
	}
	product := new(big.Int).Mul(big.NewInt(left), big.NewInt(right))
	return roundHalfAwayFromZero(product, big.NewInt(divisor))
}

// Divide calculates numerator/divisor, rounding half away from zero.
func Divide(numerator, divisor int64) (int64, error) {
	if divisor == 0 {
		return 0, ErrDivisionByZero
	}
	return roundHalfAwayFromZero(big.NewInt(numerator), big.NewInt(divisor))
}

// Clamp bounds value inclusively. The range must be ordered.
func Clamp(value, minimum, maximum int64) (int64, error) {
	if minimum > maximum {
		return 0, ErrInvalidRange
	}
	if value < minimum {
		return minimum, nil
	}
	if value > maximum {
		return maximum, nil
	}
	return value, nil
}

func roundHalfAwayFromZero(numerator, divisor *big.Int) (int64, error) {
	if divisor.Sign() == 0 {
		return 0, ErrDivisionByZero
	}

	negative := numerator.Sign() != 0 && (numerator.Sign() < 0) != (divisor.Sign() < 0)
	absoluteNumerator := new(big.Int).Abs(new(big.Int).Set(numerator))
	absoluteDivisor := new(big.Int).Abs(new(big.Int).Set(divisor))
	quotient, remainder := new(big.Int), new(big.Int)
	quotient.QuoRem(absoluteNumerator, absoluteDivisor, remainder)

	if new(big.Int).Lsh(remainder, 1).Cmp(absoluteDivisor) >= 0 {
		quotient.Add(quotient, big.NewInt(1))
	}
	if negative {
		quotient.Neg(quotient)
	}
	return checkedBigInt(quotient)
}

func checkedBigInt(value *big.Int) (int64, error) {
	if !value.IsInt64() {
		return 0, ErrOverflow
	}
	return value.Int64(), nil
}

File: server/src/market/precision.ts
/**
 * Fixed-point arithmetic for authoritative market calculations (ported from
 * market/precision). Values are scaled integers; binary floating point must
 * not be used for settlement inputs or outputs.
 */

export const WeightScale = 100_000_000;
export const ReturnScale = 100_000_000;
export const NormalizedScale = 1_000_000;
export const ScoreScale = 10_000;

const MAX_SAFE = Number.MAX_SAFE_INTEGER;

export function errDivisionByZero(): Error {
  return new Error("precision division by zero");
}

export function errOverflow(): Error {
  return new Error("precision overflow");
}

export function errInvalidRange(): Error {
  return new Error("precision invalid clamp range");
}

/** Returns left + right when the exact result fits in a safe integer. */
export function add(left: number, right: number): number {
  const result = BigInt(left) + BigInt(right);
  return checked(result);
}

/** Returns left - right when the exact result fits in a safe integer. */
export function subtract(left: number, right: number): number {
  const result = BigInt(left) - BigInt(right);
  return checked(result);
}

/** Calculates left*right/divisor, rounding half away from zero. */
export function multiplyDivide(
  left: number,
  right: number,
  divisor: number,
): number {
  if (divisor === 0) {
    throw errDivisionByZero();
  }
  const product = BigInt(left) * BigInt(right);
  return roundHalfAwayFromZero(product, BigInt(divisor));
}

/** Calculates numerator/divisor, rounding half away from zero. */
export function divide(numerator: number, divisor: number): number {
  if (divisor === 0) {
    throw errDivisionByZero();
  }
  return roundHalfAwayFromZero(BigInt(numerator), BigInt(divisor));
}

/** Bounds value inclusively; the range must be ordered. */
export function clamp(value: number, minimum: number, maximum: number): number {
  if (minimum > maximum) {
    throw errInvalidRange();
  }
  if (value < minimum) {
    return minimum;
  }
  if (value > maximum) {
    return maximum;
  }
  return value;
}

function roundHalfAwayFromZero(numerator: bigint, divisor: bigint): number {
  if (divisor === 0n) {
    throw errDivisionByZero();
  }
  const negative = numerator !== 0n && (numerator < 0n) !== (divisor < 0n);
  const absoluteNumerator = numerator < 0n ? -numerator : numerator;
  const absoluteDivisor = divisor < 0n ? -divisor : divisor;
  let quotient = absoluteNumerator / absoluteDivisor;
  const remainder = absoluteNumerator % absoluteDivisor;
  if (remainder * 2n >= absoluteDivisor) {
    quotient += 1n;
  }
  if (negative) {
    quotient = -quotient;
  }
  return checked(quotient);
}

function checked(value: bigint): number {
  if (value > BigInt(MAX_SAFE) || value < -BigInt(MAX_SAFE)) {
    throw errOverflow();
  }
  return Number(value);
}
